packages/opentelemetry-resources/src/Resource.ts

```diff
@@ -51,13 +58,14 @@ export class Resource implements IResource {
    */
   static default(): IResource {
     return new Resource({
-      [SemanticResourceAttributes.SERVICE_NAME]: defaultServiceName(),
-      [SemanticResourceAttributes.TELEMETRY_SDK_LANGUAGE]:
-        SDK_INFO[SemanticResourceAttributes.TELEMETRY_SDK_LANGUAGE],
-      [SemanticResourceAttributes.TELEMETRY_SDK_NAME]:
-        SDK_INFO[SemanticResourceAttributes.TELEMETRY_SDK_NAME],
-      [SemanticResourceAttributes.TELEMETRY_SDK_VERSION]:
-        SDK_INFO[SemanticResourceAttributes.TELEMETRY_SDK_VERSION],
+      [SEMRESATTRS_SERVICE_NAME]: defaultServiceName(),
+      [SEMRESATTRS_SERVICE_INSTANCE_ID]: randomUUID(),
```

Comment:
Hmm, IMO this is a quite substantial change.

I'd like to avoid adding these as they're still experimental and the `@opentelemetry/resource` package is stable. I'd prefer some form of opt-in mechanism for those.

It'll also likely affect metrics quite a bit as the resource is considered identifying for a metrics stream substantially increasing cardinality. 

We also always merge the resource with `Resource.default()` in SDKs, so there's no way to opt out from it.
We may have to change our SDKs first as it looks like we incorrectly implement this part of the spec:

> The SDK MUST provide access to a Resource with at least the attributes listed at
[Semantic Attributes with SDK-provided Default Value](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/README.md#semantic-attributes-with-sdk-provided-default-value).
This resource MUST be associated with a `TracerProvider` or `MeterProvider`
**if another resource was not explicitly specified**.

We, however, always merge `Resource.default()` with the provided one, which is incorrect:
- https://github.com/open-telemetry/opentelemetry-js/blob/47444f2a4c13a80f3328ffafc5b4c8db9e9a0886/packages/opentelemetry-sdk-trace-base/src/BasicTracerProvider.ts#L83
- https://github.com/open-telemetry/opentelemetry-js/blob/0f6518dd63bf86f3149437b2aaa56c880d45ef6d/packages/sdk-metrics/src/MeterProvider.ts#L49
- https://github.com/open-telemetry/opentelemetry-js/blob/65474404326d6ffbdd8e84799e88a46bfe66632a/experimental/packages/sdk-logs/src/LoggerProvider.ts#L37

I wonder if there's a conflicting spec somewhere of if we've just done it wrong the whole time. :fearful: 
Before we can proceed with this PR, we first need to sort this situation.

Reply:
Not sure if this is incorrect implementation, because the merge function uses the non-default as precedent, and only if there isn't a value it uses what the default have, meaning this part "if another resource was not explicitly specified." is working as expected, no?

And again, for the service.instance.id the use of random is only when is not set by something else, so for example, for users who want to avoid the cardinality problem, they can set the environment variable, which will take precedent. These is also a few cloud providers who set that value, so again it would be a problem for those cases.


Considering this value is part of the recommended values, I think makes sense to implemented as part of the stable package.
WDYT?

Reply:
> Not sure if this is incorrect implementation, because the merge function uses the non-default as precedent, and only if there isn't a value it uses what the default have, meaning this part "if another resource was not explicitly specified." is working as expected, no?

Hmm, I see. I think we may have different interpretations of that spec section. :thinking: 

**Walking through my thought process:**

> The SDK MUST provide access to a Resource with at least the attributes listed at
[Semantic Attributes with SDK-provided Default Value](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/README.md#semantic-attributes-with-sdk-provided-default-value).

From my understanding this is `Resource.default()`

> This resource MUST be associated with a TracerProvider or MeterProvider
if another resource was not explicitly specified.

I read this as:

> `Resource.default()` MUST be associated with a TracerProvider or MeterProvider if another resource (the one passed via the constructor in our case) was not explicitly specified.

Meaning that the one passed via the constructor should replace (not merge) with the default resource. I think Resource Attributes and Resource are two different things, and the way I read it it's actually Resource. I see this further solidified by [this part](https://github.com/open-telemetry/opentelemetry-specification/blob/f3511a5ccda376dfd1de76dfa086fc9b35b54757/specification/resource/sdk.md?plain=1#L38-L39):

> Note: This means that it is possible to **create** and **associate** a resource that
does not have all or any of the SDK-provided attributes present.

Meaning needs to be possible to **create** a resource that does not have `service.name`, `service.namespace`, `service.instance.id`, `service.telemetry.sdk`, ... (this is possible in our SDK)

But also meaning that users should be able to **associate** such a resource with an SDK. Which is currently not possible as we always merge the provided one with the defaults.

That prompted me to work my way through other language implementations:
- Java MeterProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-java/blob/001f34f1703a43cdd473d2a94d15c125f6d2d622/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProviderBuilder.java#L60) 
- Java TracerProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-java/blob/001f34f1703a43cdd473d2a94d15c125f6d2d622/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkTracerProviderBuilder.java#L68-L72)
- Java LoggerProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-java/blob/001f34f1703a43cdd473d2a94d15c125f6d2d622/sdk/logs/src/main/java/io/opentelemetry/sdk/logs/SdkLoggerProviderBuilder.java#L42-L46)
- Python MeterProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-python/blob/f56b8013f7ec5b0795f3f8945d2ebff20b15bcc3/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/__init__.py#L389-L395)
- Python TracerProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-python/blob/f56b8013f7ec5b0795f3f8945d2ebff20b15bcc3/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L385-L388)
- Python LoggerProvider: [does not merge](https://github.com/open-telemetry/opentelemetry-python/blob/f56b8013f7ec5b0795f3f8945d2ebff20b15bcc3/opentelemetry-sdk/src/opentelemetry/sdk/_logs/_internal/__init__.py#L604-L607)

I also looked at the C++ implementation but it's spread over 3 files for each implementation so it's harder to link compared to the ones above. It looks like the same is happening there.

Our challenge here is that we don't offer a way for users to choose if they want their `Resource` merged with the defaults, so if someone does not want the ones that are added in this PR, they will not have no choice about it. Especially with every metric producer suddenly becoming unique, this can get expensive quickly.

**A possible course of action:**

We cannot easily remove this default from the stable packages, as it would break the stability of their telemetry, possibly breaking dashboards and more in the process:

- We introduce a flag for each of the SDKs that allows users to opt-out of auto-merging with `Resource.default()`
- We introduce that same flag in NodeSDK so that users can set it there as well
- In SDK 2.0 we align with the other SDK implementations and the spec, where we only use `Resource.default()` when the user does NOT provide a Resource to the SDK constructor.

> Considering this value is part of the recommended values, I think makes sense to implemented as part of the stable package.
>
> WDYT?

I think I disagree with this. Adding experimental semconv attributes to a widely used SDK default puts us in difficult spot regarding breaking semconv changes. Especially as we've already marked the metrics/traces/resource packages as stable. Looking at what happened with [the HTTP semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/ccb6db6f3dbc6fdb4c2ec16b0ed8fccd1e783e4b/docs/http/migration-guide.md?plain=1#L3-L25), we're now required to double-emit telemetry with two different versions of semantic conventions, I'd like to avoid this here, as it's very hard to implement and maintain (we have not even figured out an elegant way to do this, see #4572). The impact would even be bigger than for the HTTP semconv as every SDK user would be affected.

If these attributes were already marked as stable, we'd just move forward with it as we don't have to fear breaking changes. Do we know what other SIGs do with these attributes? :thinking:

Reply:
@maryliag pinged me on this, I'll have a closer look at what the specification mandates here.

From a user point of view, however, it is beneficial to have both options: extend (merge with) the default resource, and also override (replace) the default resource. Blocking any of those two ways will block certain use cases.

.NET, for example, allows extending existing resources (via a [`ConfigureResource`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/OpenTelemetryBuilderSdkExtensions.cs#L36C41-L36C58) method on provider builders), as well as replacing the default resource (via a [`SetResourceBuilder`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/Builder/MeterProviderBuilderSdk.cs#L139) method on the provider builder). While this can probably realized in a more elegant way, the important thing is that it covers all use cases.

Reply:
> From a user point of view, however, it is beneficial to have both options: extend (merge with) the default resource, and also override (replace) the default resource. Blocking any of those two ways will block certain use cases.

Yep, I completely agree.

> .NET, for example, allows extending existing resources (via a [ConfigureResource](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/OpenTelemetryBuilderSdkExtensions.cs#L36C41-L36C58) method on provider builders), as well as replacing the default resource (via a [SetResourceBuilder](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/Builder/MeterProviderBuilderSdk.cs#L139) method on the provider builder). While this can probably realized in a more elegant way, the important thing is that it covers all use cases.

Yes, [Java implements this in a similar way](https://github.com/open-telemetry/opentelemetry-java/blob/001f34f1703a43cdd473d2a94d15c125f6d2d622/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProviderBuilder.java#L60), `setResource()` overrides, `addResource()` merges.

Maybe this is a bigger deal in my mind than it actually is:

As @maryliag correctly pointed out earlier: users can override those attributes specifically via the env provider, or via the cloud providers (they can also do it directly in code if they'd like). 

My preference is to have feature ready so that users don't have to do it every time a new default attribute is added. The easiest way to facilitate that seems to be to implement requirement from the spec. I'll prepare a PR to show how this could look like. I'll also put it on the agenda for tomorrow's SIG meeting. I'm sure we can figure something out.

Reply:
Those next steps make sense to me, thank you both for the input!

Reply:
Regarding the topic of this value being always set VS with opt-in: I used the example of how other stable SDKs deal with that (e.g. [here](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.AspNetCore#experimental-support-for-grpc-requests)), and I made a change in this PR to have the value as default as opt-in based on the env var `OTEL_NODE_JS_EXPERIMENTAL_DEFAULT_SERVICE_INSTANCE_ID`

This way this feature can still be part of the stable package, but with experimental values. And after we have that for awhile, we can change the behaviour in the future.

Then we can have another PR focusing on the merge/override, like the one @pichlermarc created on #4617